enclave/sgx/revocationinfo.c

```diff
@@ -3,161 +3,168 @@
 
 #include <openenclave/bits/safecrt.h>
 #include <openenclave/bits/safemath.h>
+#include <openenclave/corelibc/stdio.h>
 #include <openenclave/enclave.h>
 #include <openenclave/internal/calls.h>
 #include <openenclave/internal/raise.h>
 #include <stdlib.h>
 #include <string.h>
 #include "../common/sgx/revocation.h"
-
-/**
- * Validate and copy buffer to enclave memory.
- */
-static oe_result_t _copy_buffer_to_enclave(
-    uint8_t** dst,
-    size_t* dst_size,
-    const uint8_t* src,
-    size_t src_size)
-{
-    oe_result_t result = OE_FAILURE;
-    if (!src || src_size == 0 || !oe_is_outside_enclave(src, src_size) ||
-        dst == NULL || dst_size == NULL)
-        OE_RAISE(OE_INVALID_PARAMETER);
-
-    *dst = malloc(src_size);
-    if (*dst == NULL)
-        OE_RAISE(OE_OUT_OF_MEMORY);
-
-    OE_CHECK(oe_memcpy_s(*dst, src_size, src, src_size));
-    *dst_size = src_size;
-    result = OE_OK;
-
-done:
-    return result;
-}
+#include "internal_t.h"
 
 /**
  * Call into host to fetch revocation information.
  */
 oe_result_t oe_get_revocation_info(oe_get_revocation_info_args_t* args)
 {
     oe_result_t result = OE_FAILURE;
-    size_t host_args_buffer_size = sizeof(*args);
-    uint8_t* host_args_buffer = NULL;
-    oe_get_revocation_info_args_t* host_args = NULL;
-    oe_get_revocation_info_args_t tmp_args = {0};
-    uint8_t* p = NULL;
-    size_t crl_url_sizes[2] = {0};
-
-    if (args == NULL || args->num_crl_urls != 2 || args->crl_urls[0] == NULL ||
-        args->crl_urls[1] == NULL)
-        OE_RAISE(OE_FAILURE);
+    oe_get_revocation_info_args_t in = {0};
+    oe_get_revocation_info_args_t out = {0};
+    uint32_t retval;
 
-    // Compute size of buffer to allocate in host memory to marshal the
-    // arguments.
-    for (uint32_t i = 0; i < args->num_crl_urls; ++i)
-    {
-        result =
-            oe_safe_add_sizet(strlen(args->crl_urls[i]), 1, &crl_url_sizes[i]);
-        if (result != OE_OK)
-            goto done;
-
-        result = oe_safe_add_sizet(
-            host_args_buffer_size, crl_url_sizes[i], &host_args_buffer_size);
+    if (!args)
+        OE_RAISE(OE_FAILURE);
 
-        if (result != OE_OK)
-            goto done;
-    }
+    /* Verify the crl_urls. */
+    if (args->num_crl_urls != 2 || !args->crl_urls[0] || !args->crl_urls[1])
+        OE_RAISE(OE_FAILURE);
 
-    host_args_buffer = oe_host_malloc(host_args_buffer_size);
-    if (host_args_buffer == NULL)
-        OE_RAISE(OE_OUT_OF_MEMORY);
+    /* fmspc */
+    memcpy(in.fmspc, args->fmspc, sizeof(in.fmspc));
 
-    // Copy args struct.
-    p = host_args_buffer;
-    host_args = (oe_get_revocation_info_args_t*)p;
-    *host_args = *args;
-    p += sizeof(*host_args);
+    /* crl_urls */
+    memcpy(in.crl_urls, args->crl_urls, sizeof(in.crl_urls));
+    in.num_crl_urls = args->num_crl_urls;
 
-    // Copy input buffers.
-    for (uint32_t i = 0; i < args->num_crl_urls; ++i)
+    for (;;)
```

Comment:
This should be iteration capped as in `qeidinfo.c`.

Also, have we considered mitigating the perf cost of this by pre-allocating a default size that should generally work? We know the general sizes of the collateral data we're pulling for the moment; @soccerGB should have the numbers.

Reply:
Created issue #2049 and will need Cheng-Ming's input.

Reply:
@mikbras Buffer sizing aside, why do we allow this to loop infinitely as opposed to capping it as in `qeidinfo.c`